### PR TITLE
util/types: provide time Format method

### PR DESCRIPTION
**Pull request by tiancaiamao**
use Format to implement String(), so 00 month and day can be display
copy code mostly from evaluator builtinDateFormat, we can clean up them, later.

`parseFrac` overflow is irrelevant, and already in another PR, so it can be ignored when review.

### CODE REVIEW

File: util/types/format_test.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package types
+
+import (
+	"fmt"
```
Comment (hanfei1991) [inline util/types/format_test.go]: add a blank here

File: util/types/fsp.go [hunk 2]
```diff
@@ -43,25 +43,35 @@ func checkFsp(fsp int) (int, error) {
...
+		return 0, false, errors.Trace(err)
 	}
 
+	// Fill 0 when fsp > string length.
+	if fsp > len(s) {
+		v, err := strconv.ParseInt(s, 10, 64)
+		if err != nil {
+			return 0, false, errors.Trace(err)
+		}
+		return int(float64(v) * math.Pow10(fsp-len(s))), false, nil
```
Comment (hanfei1991) [inline util/types/fsp.go]: May the fsp be too long ?

File: util/types/fsp.go [hunk 3]
```diff
@@ -75,10 +85,16 @@ func parseFrac(s string, fsp int) (int, error) {
...
+	if round >= math.Pow10(fsp) {
+		// overflow
+		return 0, true, nil
+	}
+
 	// Get the final frac, with 6 digit number
 	//  0.1236 round 3 -> 124 -> 123000
 	//  0.0312 round 2 -> 3 -> 30000
-	return int(round * math.Pow10(MaxFsp-fsp)), nil
+	//  0.9999999 round 6 -> 1000000 -> overflow
```
Comment (hanfei1991) [inline util/types/fsp.go]: move this comment above line 88?

File: util/types/time.go [hunk 4]
```diff
@@ -1542,3 +1538,145 @@ func IsDateFormat(format string) bool {
...
+		fmt.Fprintf(buf, "%d", t.Time.Hour())
+	case 'h', 'I':
+		t := t.Time.Hour()
+		if t == 0 || t == 12 {
+			fmt.Fprintf(buf, "%02d", 12)
+		} else {
+			fmt.Fprintf(buf, "%02d", t%12)
+		}
+	case 'l':
+		fmt.Fprintf(buf, "%d", (t.Time.Hour()%12)+1)
```
Comment (hanfei1991) [inline util/types/time.go]: why need we mod 12 here?

File: util/types/time.go [hunk 5]
```diff
@@ -1542,3 +1538,145 @@ func IsDateFormat(format string) bool {
...
+			fmt.Fprintf(buf, "%02d:%02d:%02d PM", h-12, t.Time.Minute(), t.Time.Second())
+		}
+	case 'T':
+		fmt.Fprintf(buf, "%02d:%02d:%02d", t.Time.Hour(), t.Time.Minute(), t.Time.Second())
+	case 'S', 's':
+		fmt.Fprintf(buf, "%02d", t.Time.Second())
+	case 'f':
+		fmt.Fprintf(buf, "%06d", t.Time.Microsecond())
+	case 'U', 'u', 'V', 'v':
+		// TODO: Fix here.
```
Comment (hanfei1991) [inline util/types/time.go]: fix what?

Comment (tiancaiamao) [inline util/types/time.go]: `TODO: Fix here.` is a special comment I use, so grep can find it easily.
mysql may use Sunday or Monday as the first day of week, `U u V v` controls which.

but go always use Sunday as the first day of week, so not easy to do it now, I leave a TODO.

File: util/types/format_test.go [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+		c.Assert(err, IsNil, Commentf("parse time fail: %s", t.Input))
+		fmt.Println("lala", tm.Time.Microsecond())
+
+		str, err := tm.Format(t.Format)
+		c.Assert(err, IsNil, Commentf("time format fail: %d", i))
+		c.Assert(str, Equals, t.Expect, Commentf("no.%d \nobtain:%v \nexpect:%v\n", i,
+			str, t.Expect))
+	}
+
+	// error
```
Comment (tiancaiamao) [inline util/types/format_test.go]: there is already a TODO in the following line

File: util/types/format_test.go (hunk 1 above)
Comment (tiancaiamao) [inline util/types/format_test.go]: done

File: util/types/fsp.go (hunk 3 above)
Comment (tiancaiamao) [inline util/types/fsp.go]: this would not be included in this PR, it not diff master, but diff
tiancaiamao/timestamp-no-zero from tiancaiamao/time-format

**Comment by tiancaiamao** [comment]
PTAL @hanfei1991 @zimulala

**Comment by coocood** [comment]
LGTM

File: util/types/time_test.go [hunk 7]
```diff
@@ -322,9 +321,14 @@ func (s *testTimeSuite) getLocation(c *C) *time.Location {
 
 func (s *testTimeSuite) TestCodec(c *C) {
 	defer testleak.AfterTest(c)()
-	t, err := ParseTimestamp("2010-10-10 10:11:11")
+	// mysql timestamp value don't allow month=0 or day=0.
```
Comment (zimulala) [inline util/types/time_test.go]: s/mysql/MySQL

**Comment by zimulala** [comment]
LGTM
